reader_compare.go

```diff
@@ -0,0 +1,419 @@
+	for true {
+		next, err := HasNext(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		if !next {
+			break
+		}
+
+		ionType1 := reader1.Type()
+		ionType2 := reader2.Type()
+
+		if ionType1 != ionType2 {
+			return fmt.Errorf("expected ion types to match")
+		}
+
+		ionHashReader, ok := reader1.(*hashReader)
+		if ok && ionHashReader.isInStruct() {
+			err := CompareFieldNames(reader1, reader2)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = CompareAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareAnnotationSymbols(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareHasAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		isNull1 := reader1.IsNull()
+		isNull2 := reader2.IsNull()
+		if isNull1 != isNull2 {
+			return fmt.Errorf("expected readers to have same value for IsNull()")
+		}
+
+		switch ionType1 {
+		case ion.NullType:
+			if !isNull1 {
+				return fmt.Errorf("expected ionType1 to be null")
+			}
+			if !isNull2 {
+				return fmt.Errorf("expected ionType2 to be null")
+			}
+			break
+		case ion.BoolType, ion.IntType, ion.FloatType, ion.DecimalType, ion.TimestampType,
+			ion.StringType, ion.SymbolType, ion.BlobType, ion.ClobType:
+			err := CompareScalars(ionType1, isNull1, reader1, reader2)
+			if err != nil {
+				return err
+			}
+			break
+		case ion.StructType, ion.ListType, ion.SexpType:
+			err := reader1.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = Compare(reader1, reader2)
+			if err != nil {
+				return err
+			}
+
+			err = reader1.StepOut()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepOut()
+			if err != nil {
+				return err
+			}
+			break
+		default:
+			return &InvalidIonTypeError{ionType1}
+		}
+	}
+
+	next, err := HasNext(reader1, reader2)
+	if err != nil {
+		return err
+	}
+
+	if next {
+		return fmt.Errorf("expected HasNext() to return false")
+	}
+
+	return nil
+}
+
+func HasNext(reader1 ion.Reader, reader2 ion.Reader) (bool, error) {
+	more := reader1.Next()
+
+	if more != reader2.Next() {
+		return false, fmt.Errorf("next results don't match")
+	}
+
+	if !more {
+		if !reader1.Next() {
+			return false, fmt.Errorf("expected reader1.Next() to return true")
+		}
+		if !reader2.Next() {
+			return false, fmt.Errorf("expected reader2.Next() to return true")
+		}
+	}
+
+	return more, nil
+}
+
+func CompareFieldNames(reader1 ion.Reader, reader2 ion.Reader) error {
+	// TODO: Rework this once SymbolTokens are available
+	/*token1 := reader1.GetFieldNameSymbol()
+	token2 := reader2.GetFieldNameSymbol()
+
+	fieldName := token1.Text
+	if fieldName != token2.Text {
+		return fmt.Errorf("tokens don't match")
+	}
+
+	if fieldName != "" {
+		field1 := reader1.FieldName()
+		field2 := reader2.FieldName()
+		return CompareNonNullStrings(field1, field2, "field names don't match")
+	}*/
```

Comment:
I am not a fun of keeping commented code. Any chance temporary code can go on a branch.

Reply:
I removed the commented out code and I will create another branch for it.